### DISCUSSION TITLE
What's the most optimized way to traverse a tree? And how to test it properly?

### DISCUSSION THREAD

**Discussion by marcel0ll** [tree-sitter/tree-sitter#878] [Q&A] [answered]
Per the [docs](https://tree-sitter.github.io/tree-sitter/using-parsers#walking-trees-with-tree-cursors), it should be better to traverse the tree using a TreeCursor and walking the tree instead of recursively calling child nodes, but I am getting different results and I am not sure how to verify it.

I am working on a visitor library that allows to call a function on node enter and on node out and I am trying to do it in the best way possible. These two are the simplified versions of the code I am using:

> enter & exit are simplified in this code

## Recursive
```c
void visit_tree (TSNode node) {
  enter(node)

  unsigned child_count = ts_node_child_count(node);
  for (unsigned i = 0; i < child_count; i++) {
    TSNode child_node = ts_node_child(node, i);
    visit_tree(child_node);
  }

  exit(node)
}
```

## Cursor walk
```c
void visit_tree_cursor (TSNode tree) {
  TSTreeCursor cursor = ts_tree_cursor_new(tree);

  bool recurse = true;
  bool finished = false;
  do {
    // enter child
    if (recurse && ts_tree_cursor_goto_first_child(&cursor)) {
      recurse = true;
      node = ts_tree_cursor_current_node(&cursor);
      enter(node);
    } else {
      node = ts_tree_cursor_current_node(&cursor);
      exit(node)

      // go to sibling
      if (ts_tree_cursor_goto_next_sibling(&cursor)) {
        recurse = true;
        node = ts_tree_cursor_current_node(&cursor);
        enter(node);
      } else if (ts_tree_cursor_goto_parent(&cursor)) {
        recurse = false;
      } else {
        finished = true;
      }
    }
  } while(!finished);
  ts_tree_cursor_delete(&cursor);
}
```

When I `time` both implementations, I am getting better results of the recursive version.

I am not very experienced  at profiling C programs, but searching around I tried both `perf` and `gprof` and it seems like using the cursor api is more costly than using the node apis. Things that popped out to me were the number of calls for `ts_node_child_iterator_next` and `ts_node_new` that were higher using the walk version.

I based the code for the walk on the code from this [PR](https://github.com/tree-sitter/tree-sitter/pull/783).

Should I have implemented the walk in a different way? 

A more extensive explanation of what I am doing:
- using my visitor lib
- registering functions to be called on enter and exit of each node for code parsed using tree-sitter-javascript
- producing minified js while visiting the tree
- The file that I am mostly using to test the minifier is [phaser.js](https://github.com/photonstorm/phaser/blob/master/dist/phaser.js)

**Comment by marcel0ll** [comment, marked as answer]
> it seems like using the cursor api is more costly than using the node apis

The problem wasn't with cursor.

I believe that the efficiency problem I had was due to mixing both node and cursor apis. I may write a minimal example to test and ensure this.

For now, I rewrote the nodes enter/exit code to also use cursor and I am happy with the results. What gave me a hint about this was using `perf` with `flame graph`

One thing I noticed is that there is no `ts_tree_cursor_goto_prev_sibling`, but there is `ts_node_prev_sibling` in the api, didn't dig into why that is.

**Comment by skmendez** [comment]
For Rust, I also had this question, and so I answered it with a library: [tree-sitter-traversal](https://crates.io/crates/tree-sitter-traversal)! The library creates an `Iterator` for traversing `Trees` with a `TreeCursor`, and it's pretty efficient (zero allocations, minimal amount of cursor moves possible). It supports preorder and postorder traversal, and also works for any tree which implements a trait like the `TreeCursor`.

**Reply by dbarnett** [reply]
Neat! Seems to be broken against tree-sitter 0.21.0, but happily some nice folks picked it up to fork and publish a working version: https://crates.io/crates/baz-tree-sitter-traversal.
